File: compat/suites/dotnet-sdk/Scenario/Paths.cs
using System.Globalization;

namespace OvercastCompat.Scenario;

/// <summary>
/// Paths (compat/model/README.md § Paths): <c>$</c> is the response,
/// <c>.Name</c> selects a structure member or map key, <c>[n]</c> selects a
/// list element. Nothing else — no wildcards, filters, quoting or recursive
/// descent.
/// </summary>
/// <remarks>
/// A path is walked over the <em>document</em> form of a response
/// (<see cref="Documents"/>), not over the SDK object, so member names are the
/// modeled names every backend writes and a null property is absence rather
/// than null.
/// </remarks>
internal static class Paths
{
    /// <summary>One step of a path: a member name or a list index.</summary>
    internal readonly record struct Segment(string Member, int Index)
    {
        public bool IsIndex => Index >= 0;
    }

    /// <summary>
    /// Splits a path into its segments, rejecting anything the IR's path
    /// grammar does not admit — so a malformed path fails the step rather than
    /// silently resolving to nothing. The two are very different bugs.
    /// </summary>
    /// <exception cref="ScenarioPathException">If the path is malformed.</exception>
    public static IReadOnlyList<Segment> Parse(string path)
    {
        if (path.Length == 0 || path[0] != '$')
        {
            throw new ScenarioPathException($"path \"{path}\" does not start with $");
        }
        var segments = new List<Segment>();
        var rest = path.AsSpan(1);
        while (rest.Length > 0)
        {
            if (rest[0] == '.')
            {
                rest = rest[1..];
                var end = rest.IndexOfAny('.', '[');
                if (end < 0)
                {
                    end = rest.Length;
                }
                if (end == 0)
                {
                    throw new ScenarioPathException($"path \"{path}\" has an empty member name");
                }
                segments.Add(new Segment(rest[..end].ToString(), -1));
                rest = rest[end..];
                continue;
            }
            if (rest[0] == '[')
            {
                var end = rest.IndexOf(']');
                if (end < 0)
                {
                    throw new ScenarioPathException($"path \"{path}\" has an unterminated index");
                }
                var digits = rest[1..end].ToString();
                if (!int.TryParse(digits, NumberStyles.None, CultureInfo.InvariantCulture, out var index))
                {
                    throw new ScenarioPathException($"path \"{path}\" has a non-numeric index \"{digits}\"");
                }
                segments.Add(new Segment("", index));
                rest = rest[(end + 1)..];
                continue;
            }
            throw new ScenarioPathException($"path \"{path}\" has an unexpected character '{rest[0]}'");
        }
        return segments;
    }

    /// <summary>
    /// Walks a path over a document.
    /// </summary>
    /// <returns>
    /// False when any segment is absent — which is what <c>missing</c> tests
    /// for, and what makes an absent list count as empty for listContains and
    /// absent.
    /// </returns>
    /// <exception cref="ScenarioPathException">If the path is malformed.</exception>
    public static bool TryResolve(object? document, string path, out object? value)
    {
        value = null;
        var current = document;
        foreach (var segment in Parse(path))
        {
            if (segment.IsIndex)
            {
                if (current is not List<object?> list || segment.Index >= list.Count)
                {
                    return false;
                }
                current = list[segment.Index];
                continue;
            }
            if (current is not SortedDictionary<string, object?> members)
            {
                return false;
            }
            // The document's keys are .NET property names, which AWSSDK
            // capitalizes; the path's are the modeled member names, which are
            // not always capitalized (SQS's `queueUrls`). Bridging it here
            // rather than by rewriting paths at emit time is deliberate: a
            // failure message has to quote the path the scenario file writes,
            // or field 4 stops matching the other backends'.
            if (!members.TryGetValue(segment.Member, out current)
                && !members.TryGetValue(PropertyName(segment.Member), out current))
            {
                return false;
            }
        }
        value = current;
        return true;
    }

    /// <summary>
    /// The .NET property AWSSDK generates for a modeled member: the member name
    /// with its first letter capitalized.
    /// </summary>
    /// <remarks>
    /// Almost every AWS member is already PascalCase, but not all — SQS models
    /// ListDeadLetterSourceQueues' page as <c>queueUrls</c> and CreateQueue's
    /// tags as <c>tags</c>, and the .NET SDK spells both with a capital. Only
    /// the capitalization is reproduced: a member that needed more would not
    /// resolve and would fail the check that names it rather than passing
    /// wrongly.
    /// </remarks>
    public static string PropertyName(string member) =>
        member.Length == 0 || !char.IsLower(member[0])
            ? member
            : char.ToUpperInvariant(member[0]) + member[1..];
}

/// <summary>A path outside the IR's grammar.</summary>
internal sealed class ScenarioPathException(string message) : Exception(message);

File: compat/suites/java-sdk/src/main/java/io/overcast/compat/scenario/Paths.java
package io.overcast.compat.scenario;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Paths ({@code compat/model/README.md} § Paths): {@code $} is the response,
 * {@code .Name} selects a structure member or map key, {@code [n]} selects a
 * list element. Nothing else — no wildcards, filters, quoting or recursive
 * descent.
 *
 * <p>A path is walked over the <em>document</em> form of a response
 * ({@link Doc}), not over the SDK object, so member names are the modeled names
 * every backend writes and an unset member is absence rather than null.
 */
final class Paths {

    private Paths() {}

    /** One step of a path: a member name or a list index. */
    record Segment(String member, int index) {
        boolean isIndex() {
            return member == null;
        }
    }

    /** A resolved-or-not value. */
    record Resolved(Object value, boolean ok) {}

    /**
     * Splits a path into its segments. Anything the IR's path grammar does not
     * admit is rejected, so a malformed path fails the step rather than silently
     * resolving to nothing — the two are very different bugs.
     */
    static List<Segment> parse(String p) {
        if (p == null || p.isEmpty() || p.charAt(0) != '$') {
            throw ValueException.of("path " + Json.quote(String.valueOf(p)) + " does not start with $");
        }
        List<Segment> segs = new ArrayList<>();
        String rest = p.substring(1);
        while (!rest.isEmpty()) {
            char c = rest.charAt(0);
            if (c == '.') {
                rest = rest.substring(1);
                int end = indexOfAny(rest, ".[");
                if (end < 0) {
                    end = rest.length();
                }
                if (end == 0) {
                    throw ValueException.of("path " + Json.quote(p) + " has an empty member name");
                }
                segs.add(new Segment(rest.substring(0, end), -1));
                rest = rest.substring(end);
            } else if (c == '[') {
                int end = rest.indexOf(']');
                if (end < 0) {
                    throw ValueException.of("path " + Json.quote(p) + " has an unterminated index");
                }
                String digits = rest.substring(1, end);
                int n;
                try {
                    n = Integer.parseInt(digits);
                } catch (NumberFormatException e) {
                    n = -1;
                }
                if (n < 0) {
                    throw ValueException.of("path " + Json.quote(p) + " has a non-numeric index "
                            + Json.quote(digits));
                }
                segs.add(new Segment(null, n));
                rest = rest.substring(end + 1);
            } else {
                throw ValueException.of("path " + Json.quote(p) + " has an unexpected character "
                        + Json.quote(String.valueOf(c)));
            }
        }
        return segs;
    }

    /**
     * Walks a path over a document. {@code ok} is false when any segment is
     * absent — which is what {@code missing} tests for, and what makes an absent
     * list count as empty for {@code listContains} and {@code absent}.
     *
     * <p>A member the service sent as JSON null resolves, to {@link Json#NULL}.
     */
    static Resolved resolve(Object doc, String path) {
        Object cur = doc;
        for (Segment s : parse(path)) {
            if (s.isIndex()) {
                if (!(cur instanceof List<?> list) || s.index() >= list.size()) {
                    return new Resolved(null, false);
                }
                cur = list.get(s.index());
                continue;
            }
            if (!(cur instanceof Map<?, ?> obj)) {
                return new Resolved(null, false);
            }
            Object v = obj.get(s.member());
            if (v == null && !obj.containsKey(s.member())) {
                // The document's keys are the member names the SDK reports
                // on each SdkField, which is the modeled name — so a path
                // spelling a member the model writes lowercase (SQS models
                // ListDeadLetterSourceQueues' page as `queueUrls`) resolves
                // exactly. This retry is the tolerance behind that, for a
                // document whose key was capitalized on the way in, and it is
                // a fallback rather than a mapping: an exact key wins, and a
                // capitalized path never reaches for a lowercase one.
                String capitalized = capitalize(s.member());
                if (!obj.containsKey(capitalized)) {
                    return new Resolved(null, false);
                }
                v = obj.get(capitalized);
            }
            cur = v;
        }
        return new Resolved(cur, true);
    }

    static String capitalize(String member) {
        if (member.isEmpty() || !Character.isLowerCase(member.charAt(0))) {
            return member;
        }
        return Character.toUpperCase(member.charAt(0)) + member.substring(1);
    }

    private static int indexOfAny(String s, String chars) {
        for (int i = 0; i < s.length(); i++) {
            if (chars.indexOf(s.charAt(i)) >= 0) {
                return i;
            }
        }
        return -1;
    }
}
